format/ComputeIR.fbs

```diff
@@ -0,0 +1,210 @@
+  /// 0 refers to the left relation and 1 to the right relation.
+  relation_index: int;
+
+  /// The type of data in the referenced Field.
+  type: Type;
+}
+
+table Call {
+  /// The name of the function whose invocation this Call represents.
+  function_name: string (required);
+
+  /// Parameters for `function_name`; content/format may be unique to each
+  /// value of `function_name`.
+  options: Blob;
+
+  /// The arguments passed to `function_name`.
+  arguments: [Expression] (required);
+
+  /// The type of data which invoking `function_name` will return.
+  type: Type;
+}
+
+/// A relation is a set of rows with consitent schema.
+table Relation {
+  /// The namespaced name of this Relation.
+  ///
+  /// Names with no namespace are reserved for pure relational
+  /// algebraic operations, which currently include:
+  ///   "filter"
+  ///   "project"
+  ///   "aggregate"
+  ///   "join"
+  ///   "order_by"
+  ///   "limit"
+  ///   "literal"
+  ///   "interactive_output"
+  relation_name: string (required);
+
+  /// Parameters for `relation_name`; content/format may be unique to each
+  /// value of `relation_name`.
+  options: Blob;
+
+  /// The arguments passed to `relation_name`.
+  arguments: [Relation] (required);
+
+  /// The schema of rows in this Relation
+  schema: Schema;
+}
+
+/// The contents of Relation.options will be FilterOptions
+/// if Relation.name = "filter"
+table FilterOptions {
+  /// The expression which will be evaluated against input rows
+  /// to determine whether they should be excluded from the
+  /// "filter" relation's output.
+  filter_expression: Expression (required);
+}
+
+/// The contents of Relation.options will be ProjectOptions
+/// if Relation.name = "project"
+table ProjectOptions {
+  /// Expressions which will be evaluated to produce to
+  /// the rows of the "project" relation's output.
+  expressions: [Expression] (required);
+}
+
+/// The contents of Relation.options will be AggregateOptions
+/// if Relation.name = "aggregate"
+table AggregateOptions {
+  /// Expressions which will be evaluated to produce to
+  /// the rows of the "aggregate" relation's output.
+  aggregations: [Expression] (required);
+  /// Keys by which `aggregations` will be grouped.
+  keys: [Expression];
+}
+
+/// The contents of Relation.options will be JoinOptions
+/// if Relation.name = "join"
+table JoinOptions {
+  /// The expression which will be evaluated against rows from each
+  /// input to determine whether they should be included in the
+  /// "join" relation's output.
+  on_expression: Expression (required);
+  join_kind: string;
+}
+
+/// Whether lesser values should precede greater or vice versa.
+enum Ordering : uint8 {
+  ASCENDING,
+  DESCENDING,
+}
+
+/// Whether nulls should precede or follow other values.
+enum NullOrdering : uint8 {
+  FIRST,
+  LAST
+}
+
+table SortKey {
+  value: Expression (required);
+  ordering: Ordering = ASCENDING;
+  null_ordering: NullOrdering = LAST;
```

Comment:
Do we need to be able to control `NaN` ordering as well for floating point columns? And `NaN` versus `null` ordering?

Reply:
In general, I'd defer nan coercion to an Expression. In this instance, if `value` may have nans then it should for example be wrapped in a call to the "replace_nans_with_nulls" function to establish a consistent ordering.

Reply:
That unfortunately doesn't solve the problem. I believe there's use cases out there that need to control specific ordering with regards to `NaN` vs `null` vs other values. I.E. you could imagine all of the following being a desired output for an ascending sort:
- `[NaN, null, 1.0, 2.0, 5.0]`
- `[null, NaN, 1.0, 2.0, 5.0]`
- `[NaN, 1.0, 2.0, 5.0, null]`
- `[null, 1.0, 2.0, 5.0, NaN]`
- `[NaN, 1.0, 2.0, 5.0, null]`
- `[1.0, 2.0, 5.0, NaN, null]`
- `[1.0, 2.0, 5.0, null, NaN]`

I think to accomplish this you need something like `null_ordering`, `nan_ordering`, and `nan_vs_null` parameters.

Reply:
For example, providing a [`float_ord` function](https://gist.github.com/bkietz/8e2ef182883b886e532ffde8e537f7a3) which maps floats onto a type with less nuanced ordering could accommodate all of these cases